### PR TITLE
variable manager tab

### PR DESCRIPTION
**Pull request by jeffalo**
**Changes**

adds variable manager tab addon. ~~it is not finished yet~~.
image
image

**Reason for changes**

i saw an issue reply asking for it.

**Tests**

yep

### CODE REVIEW

**Comment by jeffalo** [comment]
Jx7kVHSWh6
latest demo

**Comment by jeffalo** [comment]
https://user-images.githubusercontent.com/40470736/108253830-82c6e400-715a-11eb-8d96-fde8de6378a9.mp4
video version

**Comment by RedGuy12** [comment]
it looks a little wide imo
________________________________

File: addons/variable-manager/userscript.js
Comment (mxmou) [inline addons/variable-manager/userscript.js:17]: Use `addon.tab.scratchClass`.

Comment (jeffalo) [reply addons/variable-manager/userscript.js:17]: thanks

**Comment by jeffalo** [comment]
i added lists, they work like variables and newlines seperate each element. press shift+enter to make a new line instead of saving

File: addons/variable-manager/userscript.js
Comment (mxmou) [reply addons/variable-manager/userscript.js]: You don't even need to use `classList`, because `scratchClass` supports multiple arguments: `addon.tab.scratchClass("gui_tab", "asset-panel_wrapper", {others: ["sa-var-manager"]})`.

Comment (jeffalo) [reply addons/variable-manager/userscript.js]: oh wow! i'll change it then.

File: addons/variable-manager/style.css
Comment (RedGuy12) [inline addons/variable-manager/style.css:10]: wouldn't it be easier to just do
`.sa-var-manager.freeze input:focus`

Comment (jeffalo) [reply addons/variable-manager/style.css:10]: oh yeah

**Comment by mxmou** [comment]
I'm not able to see the elements in devtools, it looks like they are constantly being removed and added again for some reason. This also causes textareas to reset their height when they lose focus.

**Comment by jeffalo** [comment]
> I'm not able to see the elements in devtools, it looks like they are constantly being removed and added again for some reason. This also causes textareas to reset their height when they lose focus.

yeah, that was the only way (i know of) I could do list rendering. I also update the list while it's open every vm step. that's because scratch-vm apparently doesn't have an event thing for variable update (@apple502j told me)

as for the textarea heights, i plan on removing the resize bar, and automatically setting the height based on the content

**Comment by jeffalo** [comment]
any idea on how the textarea can auto resize itself? i cant seem to find a way

**Comment by RedGuy12** [comment]
> any idea on how the textarea can auto resize itself? i cant seem to find a way

i went to the google and found this
http://jsfiddle.net/n9uuv0hd/80/

**Comment by LankyBox01** [comment]
Uhm, i tried download it, and...
uhhh help

It doesn't really work

**Comment by jeffalo** [comment]
> Uhm, i tried download it, and...
> uhhh help
> 
> It doesn't really work

you've clearly not downloaded it properly... seems like the translations haven't loaded. any errors in the console?

**Comment by RedGuy12** [comment]
> Uhm, i tried download it, and...
> uhhh help
> 
> It doesn't really work

I know the problem, disable the other scratch addons you have

**Comment by LankyBox01** [comment]
> > Uhm, i tried download it, and...
> > uhhh help
> > It doesn't really work
> 
> I know the problem, disable the other scratch addons you have

Oh, uh.. ok

**Comment by RedGuy12** [comment]
> > > Uhm, i tried download it, and...
> > > uhhh help
> > > It doesn't really work
> > 
> > 
> > I know the problem, disable the other scratch addons you have
> 
> Oh, uh.. ok

yeah you can only have one scratch addons enabled or it causes problems

**Comment by RedGuy12** [comment]
i cant push enter in the lists textarea

**Comment by jeffalo** [comment]
> i cant push enter in the lists textarea

this explains how they work https://github.com/ScratchAddons/ScratchAddons/pull/1615#issuecomment-780826177

**Comment by RedGuy12** [comment]
> > i cant push enter in the lists textarea
> 
> this explains how they work [#1615 (comment)](https://github.com/ScratchAddons/ScratchAddons/pull/1615#issuecomment-780826177)

ah ok

**Comment by jeffalo** [comment]
i highly recomend that someone who knows what they are doing rewrite this or implement some optimization. try it in a test project and it runs fine but then try appel https://scratch.mit.edu/projects/60917032/editor/

**Comment by jeffalo** [comment]
update:
performance is still wack as heck.

try the appel project, runs fine on the blank sprite. go to the player sprite. massive lag. switch to the code tab, no lag. switch to the variable tab. much lag. using browser devtools delete the `.gui_blocks-wrapper_1ccgf .box_box_2jjDp` div, works perfect.

how can we bring back the div safely? is there a way to reinitialize the blockly instance or something?

**Comment by Hans5958** [comment]
Related to #1518 (come on at least link it lol)

**Comment by jeffalo** [comment]
> Related to #1518 (come on at least link it lol)

so that's where the idea came from!!

**Comment by jeffalo** [comment]
requires optimization i do not know how to do.

**Comment by jeffalo** [comment]
garbomuffin optimzied fork: https://github.com/garbomuffin/scratchaddons/tree/variable-tab2

**Comment by Hans5958** [comment]
Just merge his branch to yours?

**Comment by Nikey9** [comment]
Thank you for directing me here @jeffalo , I didn't know y'all were already working on a variable tab!
I'm currently trying to figure out how to install the wip version that garbomuffin made to test, let's see if I find out how hahaha

**Comment by jeffalo** [comment]
> Thank you for directing me here @jeffalo , I didn't know y'all were already working on a variable tab!
> I'm currently trying to figure out how to install the wip version that garbomuffin made to test, let's see if I find out how hahaha

we are actually working on this because of your suggestion :) 

for garbomuffin's branch, try these steps https://github.com/ScratchAddons/ScratchAddons/wiki/Installing-Scratch-Addons

**Comment by Nikey9** [comment]
> > Thank you for directing me here @jeffalo , I didn't know y'all were already working on a variable tab!
> > I'm currently trying to figure out how to install the wip version that garbomuffin made to test, let's see if I find out how hahaha
> 
> we are actually working on this because of your suggestion :)
> 
> for garbomuffin's branch, try these steps https://github.com/ScratchAddons/ScratchAddons/wiki/Installing-Scratch-Addons

Awesome, thank you! I got a chance to check it out for a bit and here's the feedback I have so far (most of this is you and garbomuffin are probably aware of already but I might as well):

- It's so, so much smoother. Holy smokes. I'm super happy with how easy it is to edit text in here
- It would make more sense to give more space to the list items and less to the list name
- Having the list item numbers somewhere on there greyed out or something would help a ton
- A very light divider line between the individual list items (toggleable?) would make it easier to distinguish one list item from the next
- The absolute highlight would be a feature that lets you open multiple lists next to each other to compare them easier, but with the way this is looking right now that's probably a bit too late to request hahaha

Here's a quick screenshot of what my lists usually look like, so you can maybe see where I'm coming from with some of these requests. And again, thank you for taking the time to work on this! 
image

**Comment by jeffalo** [comment]
> Awesome, thank you! I got a chance to check it out for a bit and here's the feedback I have so far (most of this is you and garbomuffin are probably aware of already but I might as well):

thanks for the feedback, will do!

@GarboMuffin is it alright if I merge your branch into mine now?

todo list for myself

- [ ] addon description
- [ ] dark editor support
- [ ] add garbo to credits list (thank you so much once again!!)
- [ ] change widths so that the values take up 75% of the page (i don't know why i chose 25% earlier)
- [ ] perhaps some kind of optional full page editor for big lists could be really nice
- [ ] list numbers, line spacing
- [ ] list comparisons might be tricky, i guess i can try but i don't know if it's worth it
- [ ] perhaps since this is a big addon, some way to link to the feedback form while it's in beta cc @WorldLanguages, is that okay?

**Comment by GarboMuffin** [comment]
> @GarboMuffin is it alright if I merge your branch into mine now?

yes

**Comment by TheColaber** [comment]
>  dark editor support

This can be easily done by 
> use classes instead of styling all tables (not sure if scratch-gui uses any but better safe than sorry)

because this is what we are aiming for... Not to use any of our own classes in the editor when possible.

Also this:
> perhaps since this is a big addon, some way to link to the feedback form while it's in beta cc @WorldLanguages, is that okay?

I'll have the beta testers focus on this addon :)

**Comment by jeffalo** [comment]
before this can get merged, theres a bug with `addon.tab.scratchClass`. if you are not on the page, it won't be able to find the class and it'll return an invaid class. perhaps it needs to return a promise which can be awaited. cc @WorldLanguages

**Comment by jeffalo** [comment]
i switched to using a normal querySelector which fixes it for this addon.

File: addons/variable-manager/userscript.js
Comment (WorldLanguages) [inline addons/variable-manager/userscript.js]: Calls to `url()` in non-theme CSS (like this case) are relative to the CSS file's URL, so you don't need to do this from JS.

**Comment by jeffalo** [comment]
i did not add a link to the feedback form since it's probably not fun for the translators to do.

File: addons/variable-manager/userscript.js
Comment (GarboMuffin) [inline addons/variable-manager/userscript.js:192]:
```suggestion
    localVariables = editingTarget.isStage ? [] : Object.values(editingTarget.variables).map((i) => new WrappedVariable(i, editingTarget));
```

Comment (jeffalo) [reply addons/variable-manager/userscript.js]: thanks

**Comment by jeffalo** [comment]
wait one minor thing before we merge it'll will bug me until i fix it.
image

will fix after i quickly eat dinner 😅

File: addons/variable-manager/userscript.js
Comment (GarboMuffin) [inline addons/variable-manager/userscript.js]: dont

Comment (jeffalo) [reply addons/variable-manager/userscript.js]: oh

**Comment by jeffalo** [comment]
ready

**Review by WorldLanguages — APPROVED** [review]

File: addons/variable-manager/style.css
Comment (RedGuy12) [inline addons/variable-manager/style.css:111]: why commented out?

**Review by RedGuy12 — CHANGES_REQUESTED** [review]

File: addons/variable-manager/style.css
Comment (mxmou) [inline addons/variable-manager/style.css:102]: `.sa-var-manager th`?

File: addons-l10n/en/variable-manager.json
Comment (WorldLanguages) [reply addons-l10n/en/variable-manager.json:3]: what's the grammar issue here?

Comment (RedGuy12) [reply addons-l10n/en/variable-manager.json:3]: capitalization

Comment (WorldLanguages) [reply addons-l10n/en/variable-manager.json:3]: eh, questionable., up to @jeffalo

Comment (mxmou) [reply addons-l10n/en/variable-manager.json:3]: @WorldLanguages I think addons that add UI to Scratch should use capitalization the same way as Scratch. Only context menu items don't need to be capitalized.

File: addons/variable-manager/style.css
Comment (jeffalo) [reply addons/variable-manager/style.css:111]: because it's something i might reconsider one day?

File: addons/variable-manager/style.css
Comment (jeffalo) [reply addons/variable-manager/style.css:50]: this is better than adding the same class to every element imo. it's ugly but atleast it works.

Comment (mxmou) [reply addons/variable-manager/style.css:50]: @jeffalo [Beautiful is better than ugly.](https://www.python.org/dev/peps/pep-0020/#:~:text=Beautiful%20is%20better%20than%20ugly.) I don't see what's wrong with adding a class to every element.

Comment (jeffalo) [reply addons/variable-manager/style.css:50]: alright then. i think the main issue was figuring out how to get everything except the focused input to be greyed out. i don't know if i can do it in pure CSS so maybe I'll need to use js for adding the classes and handling all that. ill have time tomorrow to try

File: addons-l10n/en/variable-manager.json
Comment (WorldLanguages) [reply addons-l10n/en/variable-manager.json:3]: Good point, let's capitalize then?

File: addons/variable-manager/style.css
Comment (mxmou) [reply addons/variable-manager/style.css:50]: > i think the main issue was figuring out how to get everything except the focused input to be greyed out. i don't know if i can do it in pure CSS

Something like this might work (but I haven't tested): `.sa-var-manager table:focus-within input:not(:focus) {color: gray;}`

**Comment by Nikey9** [comment]
Heya! Where can I find the newest version of this? I tried downloading garbomuffins fork, but it seems like that's still the old one or maybe I'm mistaken.

**Comment by RedGuy12** [comment]
> Heya! Where can I find the newest version of this? I tried downloading garbomuffins fork, but it seems like that's still the old one or maybe I'm mistaken.

@Nikey9 https://github.com/jeffalo/ScratchAddons/tree/variable-tab

**Comment by jeffalo** [comment]
todo list for myself.
- [ ] capitalisation
- [ ] fix that one css selector
- [ ] experiment with trying to fix the really long selector

File: addons/variable-manager/userscript.js
Comment (Hans5958) [inline addons/variable-manager/userscript.js:40]: Is it possible to add the version number on it, just in case?

Optional: Is it possible to use flex so the feedback text is exactly on the right of the search box?

**Review by Hans5958 — CHANGES_REQUESTED** [review]

File: addons/variable-manager/userscript.js
Comment (jeffalo) [reply addons/variable-manager/userscript.js:40]: i believe w_l said on the discord that addons don't know the SA version number but i guess it wouldn't be too hard to implement it into the addon api.

as for the feedback text, i'm not really too concerned about it because it's probably going to be removed in a few weeks. but i'll experiment/

**Comment by mxmou** [comment]
I fixed the bottom border of the tab being darker than that of the other tabs. Can I add the commit now or would that cause problems?

**Comment by jeffalo** [comment]
> I fixed the bottom border of the tab being darker than that of the other tabs. Can I add the commit now or would that cause problems?

go for it

File: addons/variable-manager/userscript.js
Comment (Hans5958) [reply addons/variable-manager/userscript.js:40]: > as for the feedback text, i'm not really too concerned about it because it's probably going to be removed in a few weeks. but i'll experiment/

I mean we got two weeks minimum before the beta tag is removed, so yeah.

**Comment by jeffalo** [comment]
ok it's ready for real this time i am sorry it's been taking so long

**Review by Hans5958 — APPROVED** [review]
I should have approved this because I have tested it and it works, and those changes are CSS only.

**Comment by Hans5958** [comment]
I'm confused, why do the checks didn't work here?

**Comment by Genius398** [comment]
>  GjeGY7DUSk.mp4 
> video version

Please add some CSS in it.

**Comment by jeffalo** [comment]
> > GjeGY7DUSk.mp4
> > video version
> 
> Please add some CSS in it.

that was the first video since then it's changed a lot

**Comment by superscratch444tre** [comment]
image

that text should be white in dark mode

**Comment by mxmou** [comment]
> image
> 
> that text should be white in dark mode

I think it's intentionally gray.

**Comment by Genius398** [comment]
> > > GjeGY7DUSk.mp4
> > > video version
> > 
> > 
> > Please add some CSS in it.
> 
> that was the first video since then it's changed a little bit. and what do you mean by some CSS? it already had some

ok

**Comment by superscratch444tre** [comment]
> > image
> > 
> > that text should be white in dark mode
> 
> I think it's intentionally gray.

Since in normal mode that text is dark black, it should be white in dark mode.

**Comment by mxmou** [comment]
> Since in normal mode that text is black, it should be white in dark mode.

In normal mode it's dark gray.

**Comment by jeffalo** [comment]
@superscratch444tre its really fine, it's not meant to stand out much

**Pull request merged**
